File: 1717-maximum-score-from-removing-substrings/1717-maximum-score-from-removing-substrings.py
class Solution:
    def solve(self, s: str, targetStr: str, points: int) -> tuple[int, str]:
        totalPoints = 0
        n = len(s)
        write_idx = 0
        s = list(s)
        for read_idx in range(n):
            s[write_idx] = s[read_idx]
            write_idx += 1
            if write_idx > 1 and s[write_idx - 1] == targetStr[1] and s[write_idx - 2] == targetStr[0]:
                write_idx -= 2
                totalPoints += points
        
        s = ''.join(s[:write_idx])
        return totalPoints, s

    def maximumGain(self, s: str, x: int, y: int) -> int:
        s1, s2 = "ab", "ba"
        if y > x:
            s1, s2 = s2, s1
            x, y = y, x

        points1, s = self.solve(s, s1, x)
        points2, _ = self.solve(s, s2, y)

        return points1 + points2





# class Solution:
#     def solve(self, s: str, targetStr: str, points: int) -> int:
#         totalPoints = 0
#         st = []
#         for char in s:
#             if st and char == targetStr[1] and st[-1] == targetStr[0]:
#                 st.pop()
#                 totalPoints += points
#             else:
#                 st.append(char)
        
#         s = ''.join(st)
#         return totalPoints, s

#     def maximumGain(self, s: str, x: int, y: int) -> int:
#         s1, s2 = "ab", "ba"
#         if y > x:
#             x, y = y, x
#             s1, s2 = s2, s1
        
#         points1, s = self.solve(s, s1, x)
#         points2, _ = self.solve(s, s2, y)
        
#         return points1 + points2

File: LeetCode/1717-maximum-score-from-removing-substrings/1717-maximum-score-from-removing-substrings.cpp
class Solution {
public:
    int solve(string &s, string target, int point) {
        int totalPoints = 0;
        int n = s.length();
        int writeidx = 0;
        for (int readidx = 0; readidx < n; readidx++) {
            s[writeidx++] = s[readidx];
            if (writeidx > 1 && s[writeidx-1] == target[1] && s[writeidx-2] == target[0]) {
                writeidx -= 2;
                totalPoints += point;
            }
        }
        s = s.substr(0, writeidx);
        return totalPoints;
    }
    int maximumGain(string s, int x, int y) {
        string s1 = "ab", s2 = "ba";
        if (y > x) {
            swap(x, y);
            swap(s1, s2);
        }
        return solve(s, s1, x) + solve(s, s2, y);
    }
};





// class Solution {
// public:
//     int solve(string &s, string targetStr, int point) {
//         int totalPoints = 0, n = s.length();
//         stack<char> st;
//         for (int i = 0; i < n; i++) {
//             if (st.empty()) {
//                 st.push(s[i]);
//             } else {
//                 if (s[i] == targetStr[1] && st.top() == targetStr[0]) {
//                     st.pop();
//                     totalPoints += point;
//                 } else {
//                     st.push(s[i]);
//                 }
//             }
//         }
//         s = "";
//         while (!st.empty()) {
//             s += st.top();
//             st.pop();
//         }
//         reverse(s.begin(), s.end());
//         return totalPoints;
//     }
//     int maximumGain(string s, int x, int y) {
//         string s1 = "ab" , s2 = "ba";
//         if (x < y) {
//             swap(s1, s2);
//             swap(x, y);
//         }
//         return solve(s, s1, x) + solve(s, s2, y);
//     }
// };